### PR TITLE
add TalkProposal export columns "cancelled"、"accepted"、"vote"

### PR DESCRIPTION
**Pull request by Maliaotw**
(Thanks for sending a pull request! Please fill in the following content to let us know better about this change.)

## Types of changes
Please put an `x` in the box that applies

- [ ] **Bugfix**
- [x] **New feature**
- [ ] **Refactoring**
- [ ] **Breaking change** (any change that would cause existing functionality to not work as expected)
- [ ] **Documentation Update**
- [ ] **Other (please describe)**

## Description
追加admin下的Proposal表，多匯出幾個字段
#779 加入兩個字段匯出"canceled" and "is_accepted"
#780 加入關聯review表下的vote字段匯出

## Steps to Test This Pull Request
1. Go to 'admin/proposals/talkproposal'
2. Click on 'list_display'
3. Click on to 'export'

### CODE REVIEW

File: src/proposals/admin.py [hunk 1]
```diff
@@ -16,9 +16,13 @@ class AdditionalSpeakerInline(GenericTabularInline):
...
 
 class ProposalAdmin(admin.ModelAdmin):
+
+    def vote(self,obj):
+        return sum([int(review.vote) for review in obj.review_set.all()])
+
     list_display = [
         'title', 'category', 'duration', 'language',
-        'python_level', 'remoting_policy', 'accepted',
+        'python_level', 'remoting_policy', 'cancelled','accepted','vote'
```
Comment (Lee-W) [inline src/proposals/admin.py:25]: Hi, thanks for your contribution! 🎉
Are these the columns that will be exported? If so, #779 looks good to me.

As for #780, we don't need `vote` column, instead, we'll need 8 more columns which are `stage_1_plus_1_count`, `stage_1_plus_0_count`, `stage_1_minus_0_count`, `stage_1_minus_1_count`, `stage_2_plus_1_count`, `stage_2_plus_0_count`, `stage_2_minus_0_count`, `stage_2_minus_1_count`. Feel free to rename these columns.

Comment (Maliaotw) [reply src/proposals/admin.py]: 我嘗試修改了，請再看看，不知道這樣對不對。

**Comment by tai271828** [comment]
@Lee-W I have deployed this PR (merged with master) for you to verify. May you have a look of http://staging.pycon.tw/ to have a try if the result meets your expectation?

**Comment by Lee-W** [comment]
> @Lee-W I have deployed this PR (merged with master) for you to verify. May you have a look of http://staging.pycon.tw/ to have a try if the result meets your expectation?

It seems I do not have the admin account on stage to do the test

File: src/proposals/admin.py (hunk 1 above)
Comment (Lee-W) [reply src/proposals/admin.py]: 以原本開的需求是沒錯的

但我突然想到還有另一個情況要加ＱＱ （不過這個可以先進）
就是如果一個 reviewer 在 stage 1, stage 2 都有給 review ，我們只需要採計 stage 2 的

Comment (Lee-W) [reply src/proposals/admin.py]: 另外能幫我加一個 last_update_time （我記得有存一個類似的） 的欄位嗎

這個目的是在於讓議程組分辨這個提案在修改階段有沒有進行修改
雖然能回傳 last_update_time 直接跟修改階段結束時間比較的 bool 會比較好，但好像會有點麻煩
可以先丟 last_update_time 就好

**Comment by tai271828** [comment]
> > @Lee-W I have deployed this PR (merged with master) for you to verify. May you have a look of http://staging.pycon.tw/ to have a try if the result meets your expectation?
> 
> It seems I do not have the admin account on stage to do the test

@yychen   do you know how to grant @Lee-W  to get the permission? I don't think he needs an admin account. By adding him into some certain group should work, right? (but I don't know which group it is)

**Comment by yychen** [comment]
> 
> @yychen do you know how to grant @Lee-W to get the permission? I don't think he needs an admin account. By adding him into some certain group should work, right? (but I don't know which group it is)
> 
> @Lee-W before @yychen 's reply, let me send you the superadmin credential in private.

I believe it is being taken care of. I've contacted him this morning. Sorry for the mess.

**Comment by tai271828** [comment]
@Maliaotw   CI failed for


```
$ python src/manage.py migrate --verbosity 1 --noinput
SystemCheckError: System check identified some issues:
ERRORS:
<class 'proposals.admin.TalkProposalAdmin'>: (admin.E006) The value of 'fields' contains duplicate field(s).
<class 'proposals.admin.TutorialProposalAdmin'>: (admin.E006) The value of 'fields' contains duplicate field(s).
```

May you update the corresponding test cases?

**Comment by yychen** [comment]
As per @Lee-W 's request, there are a few columns that need to be added to the export file:
* last_updated_at
* cancelled
* remoting_policy
* referring_policy

And, here's a tricky one. For the `stage_XXXX_count` fields, we need to exclude the counts from the same reviewer. For example, there might be a reviewer that gave a review with minus 0 at stage 1, and then another review is given with plus 1 at stage 2. Only the latter one should be taken into account.

For the ease of development, I'm thinking about building a simple map. Something like adding a function to the `TalkProposalResource` class:

```python
def prepare(self, obj):
    self.count = 0
    self.data = {
        'stage_1': {"+1": 0, "+0": 0, "-0": 0, "-1": 0},
        'stage_2': {"+1": 0, "+0": 0, "-0": 0, "-1": 0},
    }
    reviewer = []
    for review in obj.review_set.all().order_by('-updated'):
        if review.reviewer.email not in reviewer:
            reviewer.append(review.reviewer.email)
            self.data['stage_%d' % review.stage][review.vote] += 1
            self.count += 1
```
You only call prepare once, and in those `dehydrate_stage_XXXX_count` functions, do something like:
```python
return self.data['stage_1']['+1']
```

Maybe take a look at https://django-import-export.readthedocs.io/en/latest/api_resources.html#modelresource and see if there's a proper function to put `prepare`. This is just a quick implementation of the requirement, maybe not the ideal or best one.

Any thoughts @tai271828 @uranusjr ?

**Comment by Maliaotw** [comment]
@yychen 
我先做了last_updated_at這個字段的匯出，及嘗試使用prepare方法進行預處理跟匯出，但我不太明白reviewer陣列之後要怎麼使用，另外還有其他沒做到的地方嗎？

**Comment by yychen** [comment]
> 
> 
> @yychen
> 我先做了last_updated_at這個字段的匯出，及嘗試使用prepare方法進行預處理跟匯出，但我不太明白reviewer陣列之後要怎麼使用，另外還有其他沒做到的地方嗎？

目前的修改看起來不錯。reviewer 陣列之後要怎麼使用的意思是什麼? 不太明白。

下面這三個欄位也要
- cancelled
- remoting_policy
- referring_policy

**Comment by Maliaotw** [comment]
> > @yychen
> > 我先做了last_updated_at這個字段的匯出，及嘗試使用prepare方法進行預處理跟匯出，但我不太明白reviewer陣列之後要怎麼使用，另外還有其他沒做到的地方嗎？
> 
> 目前的修改看起來不錯。reviewer 陣列之後要怎麼使用的意思是什麼? 不太明白。
> 
> 下面這三個欄位也要
> 
> * cancelled
> * remoting_policy
> * referring_policy

reviewer陣列問題，之前沒看明白以為沒用到，其他三個欄位都更新好了。

File: src/proposals/resources.py [hunk 2]
```diff
@@ -1,17 +1,70 @@
...
 
     class Meta:
         model = TalkProposal
         fields = [
             'id', 'title', 'category', 'python_level', 'duration',
-            'language', 'name', 'email',
+            'language', 'name', 'email', 'cancelled', 'accepted', 'last_updated_at',
+            'stage_1_plus_1_count', 'stage_1_plus_0_count', 'stage_1_minus_0_count', 'stage_1_minus_1_count',
+            'stage_2_plus_1_count', 'stage_2_plus_0_count', 'stage_2_minus_0_count', 'stage_2_minus_1_count',
+            'cancelled', 'remoting_policy', 'referring_policy'
```
Comment (yychen) [inline src/proposals/resources.py:68]: 這邊 cancelled 重複了... 保留一個就好

File: src/proposals/resources.py [hunk 3]
```diff
@@ -1,17 +1,70 @@
...
+        for review in obj.review_set.all().order_by('-updated'):
+            if review.reviewer.email not in reviewer:
+                reviewer.append(review.reviewer.email)
+                self.data['stage_%d' % review.stage][review.vote] += 1
+                self.count += 1
+
+    def before_export(self, queryset, *args, **kwargs):
+        super().before_export(queryset, *args, **kwargs)
+        queryset = self.get_queryset()
+        list(map(lambda obj: self.prepare(obj), queryset))
```
Comment (yychen) [inline src/proposals/resources.py:59]: 這邊這樣寫會有問題... 之前應該沒仔細看到，剛剛跑出來的結果是錯的。
原因是 `before_export` 是在整個表單產生之前跑的，但其實當初 `prepare` 的設計是一個 row 之前跑一次。
但看來這個套件沒有這樣的支援處理，因此簡單的解法就是，把 obj.id 也存進去。

類似下面這樣:
在 `before_export` 裡面
```python
self.data = {}
```

然後在 `prepare` 裡面使用 self.data 都要變成
```python
self.data[obj.id]...
```

接著上面的 `dehydrate_stage_...` 的 function 也每一個都要用 `self.data[obj.id][...][...]` 這樣

Comment (Maliaotw) [reply src/proposals/resources.py]: 感謝Tom指導，嘗試修改了，不知道這樣對不對

**Pull request merged**